File: engine/crates/shared/vfs/game_paths.rs
//! Game paths generation and validation.
//!
//! This module provides cross-platform game path management.
//! All path generation logic is centralized here to ensure consistency
//! across different platforms (Android, iOS, Desktop).

use std::path::{Path, PathBuf};

/// Error types for game path operations.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GamePathError {
    /// Game ID is invalid (wrong format or length)
    InvalidGameId(String),
    /// Base directory is invalid
    InvalidBaseDir,
}

impl std::fmt::Display for GamePathError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            GamePathError::InvalidGameId(id) => {
                write!(f, "Invalid game ID '{}': must be 1-64 alphanumeric, underscore or hyphen", id)
            }
            GamePathError::InvalidBaseDir => write!(f, "Invalid base directory"),
        }
    }
}

impl std::error::Error for GamePathError {}

/// Validate a game ID.
///
/// # Rules
/// - Length: 1-64 characters
/// - Allowed characters: a-z, A-Z, 0-9, _, -
///
/// # Examples
/// ```
/// use shared::vfs::game_paths::validate_game_id;
///
/// assert!(validate_game_id("my-game").is_ok());
/// assert!(validate_game_id("game_123").is_ok());
/// assert!(validate_game_id("").is_err());
/// assert!(validate_game_id("../hack").is_err());
/// ```
pub fn validate_game_id(game_id: &str) -> Result<(), GamePathError> {
    // Hand-rolled validation replaces regex crate dependency for this single use.
    // Allowed: a-z, A-Z, 0-9, underscore, hyphen. Length: 1-64.
    if game_id.is_empty() || game_id.len() > 64 {
        return Err(GamePathError::InvalidGameId(game_id.to_string()));
    }
    for b in game_id.bytes() {
        match b {
            b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'-' => {}
            _ => return Err(GamePathError::InvalidGameId(game_id.to_string())),
        }
    }
    Ok(())
}

/// Paths for a specific game instance.
///
/// All paths are absolute and isolated per game.
/// This struct is created by the runtime and can be serialized
/// back to the platform layer for cleanup operations.
#[derive(Debug, Clone)]
pub struct GamePaths {
    /// Game identifier
    game_id: String,
    /// Game code directory (read-only for the game)
    code_dir: PathBuf,
    /// User data directory (persistent, survives app updates)
    user_data_dir: PathBuf,
    /// Cache directory (may be cleared by system)
    cache_dir: PathBuf,
    /// Temporary directory (cleared when session ends)
    temp_dir: PathBuf,
}

impl GamePaths {
    /// Base subdirectory for all games.
    const GAMES_SUBDIR: &'static str = "migo/games";

    /// Create game paths from base directories and game ID.
    ///
    /// # Arguments
    /// * `files_dir` - Platform's persistent files directory
    ///   - Android: `Context.getFilesDir()`
    ///   - iOS: `NSDocumentDirectory`
    ///   - Desktop: `~/.local/share/app/` or `%APPDATA%/app/`
    /// * `cache_dir` - Platform's cache directory
    ///   - Android: `Context.getCacheDir()`
    ///   - iOS: `NSCachesDirectory`
    ///   - Desktop: `~/.cache/app/` or `%TEMP%/app/`
    /// * `game_id` - Unique game identifier
    ///
    /// # Returns
    /// `GamePaths` on success, `GamePathError` if game_id is invalid.
    pub fn new(
        files_dir: impl AsRef<Path>,
        cache_dir: impl AsRef<Path>,
        game_id: &str,
    ) -> Result<Self, GamePathError> {
        validate_game_id(game_id)?;

        let files_base = files_dir.as_ref().join(Self::GAMES_SUBDIR).join(game_id);
        let cache_base = cache_dir.as_ref().join(Self::GAMES_SUBDIR).join(game_id);

        Ok(Self {
            game_id: game_id.to_string(),
            code_dir: files_base.join("code"),
            user_data_dir: files_base.join("user_data"),
            cache_dir: cache_base.clone(),
            temp_dir: cache_base.join("tmp"),
        })
    }

    /// Get the game ID.
    pub fn game_id(&self) -> &str {
        &self.game_id
    }

    /// Get the code directory path (read-only for game).
    pub fn code_dir(&self) -> &Path {
        &self.code_dir
    }

    /// Get the user data directory path (persistent).
    pub fn user_data_dir(&self) -> &Path {
        &self.user_data_dir
    }

    /// Get the cache directory path.
    pub fn cache_dir(&self) -> &Path {
        &self.cache_dir
    }

    /// Get the temporary directory path.
    pub fn temp_dir(&self) -> &Path {
        &self.temp_dir
    }

    /// Ensure all directories exist.
    ///
    /// Creates the directories if they don't exist.
    /// Note: `code_dir` is created by the game deployment process,
    /// not by this method.
    pub fn ensure_directories(&self) -> std::io::Result<()> {
        std::fs::create_dir_all(&self.user_data_dir)?;
        std::fs::create_dir_all(&self.cache_dir)?;
        std::fs::create_dir_all(&self.temp_dir)?;
        Ok(())
    }

    /// Clean the temporary directory.
    ///
    /// Removes all files in the temp directory but keeps the directory itself.
    pub fn clean_temp(&self) -> std::io::Result<()> {
        if self.temp_dir.exists() {
            std::fs::remove_dir_all(&self.temp_dir)?;
        }
        std::fs::create_dir_all(&self.temp_dir)?;
        Ok(())
    }

    /// Clean the cache directory (including temp).
    ///
    /// Removes all cached files. User data is preserved.
    pub fn clean_cache(&self) -> std::io::Result<()> {
        if self.cache_dir.exists() {
            std::fs::remove_dir_all(&self.cache_dir)?;
        }
        std::fs::create_dir_all(&self.cache_dir)?;
        std::fs::create_dir_all(&self.temp_dir)?;
        Ok(())
    }

    /// Delete all game data (uninstall).
    ///
    /// Removes both persistent and cache data.
    /// Call this when the user uninstalls a game.
    pub fn delete_all(&self) -> std::io::Result<()> {
        // Delete files dir (user_data + code)
        let files_game_dir = self.user_data_dir.parent().unwrap_or(&self.user_data_dir);
        if files_game_dir.exists() {
            std::fs::remove_dir_all(files_game_dir)?;
        }

        // Delete cache dir
        if self.cache_dir.exists() {
            std::fs::remove_dir_all(&self.cache_dir)?;
        }

        Ok(())
    }

    /// Get the total size of user data in bytes.
    pub fn user_data_size(&self) -> u64 {
        dir_size(&self.user_data_dir)
    }

    /// Get the total size of cache in bytes.
    pub fn cache_size(&self) -> u64 {
        dir_size(&self.cache_dir)
    }

    /// Serialize paths to a format suitable for FFI/JNI.
    ///
    /// Returns paths as strings for cross-language communication.
    pub fn to_strings(&self) -> GamePathStrings {
        GamePathStrings {
            game_id: self.game_id.clone(),
            code_dir: self.code_dir.to_string_lossy().into_owned(),
            user_data_dir: self.user_data_dir.to_string_lossy().into_owned(),
            cache_dir: self.cache_dir.to_string_lossy().into_owned(),
            temp_dir: self.temp_dir.to_string_lossy().into_owned(),
        }
    }
}

/// String representation of game paths for FFI.
#[derive(Debug, Clone)]
pub struct GamePathStrings {
    pub game_id: String,
    pub code_dir: String,
    pub user_data_dir: String,
    pub cache_dir: String,
    pub temp_dir: String,
}

/// Calculate directory size recursively.
fn dir_size(path: &Path) -> u64 {
    if !path.exists() {
        return 0;
    }

    let mut size = 0u64;
    if let Ok(entries) = std::fs::read_dir(path) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                size += entry.metadata().map(|m| m.len()).unwrap_or(0);
            } else if path.is_dir() {
                size += dir_size(&path);
            }
        }
    }
    size
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_game_ids() {
        assert!(validate_game_id("my-game").is_ok());
        assert!(validate_game_id("game_123").is_ok());
        assert!(validate_game_id("GAME").is_ok());
        assert!(validate_game_id("a").is_ok());
        assert!(validate_game_id("a".repeat(64).as_str()).is_ok());
    }

    #[test]
    fn test_invalid_game_ids() {
        assert!(validate_game_id("").is_err());
        assert!(validate_game_id("a".repeat(65).as_str()).is_err());
        assert!(validate_game_id("../hack").is_err());
        assert!(validate_game_id("game/sub").is_err());
        assert!(validate_game_id("game.name").is_err());
        assert!(validate_game_id("game name").is_err());
    }

    #[test]
    fn test_game_paths_creation() {
        let paths = GamePaths::new("/data/files", "/data/cache", "test-game").unwrap();
        
        assert_eq!(paths.game_id(), "test-game");
        assert!(paths.code_dir().ends_with("migo/games/test-game/code"));
        assert!(paths.user_data_dir().ends_with("migo/games/test-game/user_data"));
        assert!(paths.cache_dir().ends_with("migo/games/test-game"));
        assert!(paths.temp_dir().ends_with("migo/games/test-game/tmp"));
    }
}

File: platforms/android/library/src/main/java/com/migo/runtime/GamePaths.java
package com.migo.runtime;

import java.io.File;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.function.IntPredicate;
import java.util.regex.Pattern;

/**
 * Manages isolated file paths for a specific game.
 * <p>
 * Each game has its own sandboxed directories:
 * <ul>
 *   <li>{@code /user} → User data (saves, preferences) - persistent</li>
 *   <li>{@code /cache} → Cache files - may be cleared by system</li>
 *   <li>{@code /code} → Game code - read only</li>
 *   <li>{@code /tmp} → Temporary files - private to one session, cleared when
 *       that session ends</li>
 * </ul>
 * <p>
 * {@link #getCacheDir()} is the per-game cache root rather than the directory
 * {@code /cache} resolves to; see its documentation.
 * <p>
 * Every path here is derived the same way as
 * {@code shared::vfs::game_paths::GamePaths} in the engine, and the two must move
 * together. They are two computations of one layout: the engine resolves
 * {@code /tmp} for the content that writes it, and this class is what deletes it,
 * so a change on one side alone moves what is written without moving what is
 * removed — or the reverse.
 *
 */
public final class GamePaths {

    // Game ID validation: lower-case alphanumeric, underscore, hyphen, 1-64
    // chars. This rule is `shared::vfs::game_paths::validate_game_id` in the
    // engine, and the two must move together: the engine refuses a session
    // whose id this class accepted. Both sides are pinned by the shared table
    // in engine/crates/shared/src/vfs/game-id-vectors.txt.
    private static final Pattern GAME_ID_PATTERN = Pattern.compile("^[a-z0-9_-]{1,64}$");

    /**
     * Ids that name a device rather than a file on Windows.
     *
     * <p>Refused here too, on Android, so that one id is valid on every
     * platform the runtime targets or on none of them. A host that ships the
     * same catalogue to a Windows build would otherwise find these titles
     * unable to create their storage directory there and nowhere else.
     */
    private static final Set<String> RESERVED_GAME_IDS = reservedGameIds();

    private static Set<String> reservedGameIds() {
        Set<String> reserved = new HashSet<>();
        reserved.add("con");
        reserved.add("prn");
        reserved.add("aux");
        reserved.add("nul");
        for (char digit = '0'; digit <= '9'; digit++) {
            reserved.add("com" + digit);
            reserved.add("lpt" + digit);
        }
        return Collections.unmodifiableSet(reserved);
    }

    private static final String MIGO_GAMES_DIR = "migo/games";

    /** Parent of the per-session temporary directories. */
    private static final String TEMP_DIR = "tmp";

    private final String gameId;
    private final File userDataDir;
    private final File cacheDir;
    private final File codeDir;
    private final File tempRoot;
    private final File tempDir;

    /**
     * Create game paths from RuntimeConfig, game ID and session ID.
     *
     * @param config    RuntimeConfig with base directories
     * @param gameId    Unique game identifier
     * @param sessionId The session these paths belong to, which is what keeps
     *                  {@code /tmp} private to it. Two concurrently live sessions
     *                  of the same game share {@code /user} and {@code /cache} on
     *                  purpose — that is one game's save data — but {@code /tmp}
     *                  lasts for a session, and sharing it made the first session
     *                  to close delete the other's live files mid-write.
     * @throws IllegalArgumentException if gameId is invalid
     */
    public GamePaths(RuntimeConfig config, String gameId, int sessionId) {
        if (!isValidGameId(gameId)) {
            throw new IllegalArgumentException(
                "Invalid gameId: must be 1-64 lower-case alphanumeric characters, underscore or hyphen, and not a reserved device name");
        }
        this.gameId = gameId;

        // Persistent storage: filesDir/migo/games/{gameId}/
        File filesBase = new File(config.getFilesDir(), MIGO_GAMES_DIR + "/" + gameId);
        this.userDataDir = new File(filesBase, "user_data");
        this.codeDir = new File(filesBase, "code");

        // Cache storage: cacheDir/migo/games/{gameId}/
        File cacheBase = new File(config.getCacheDir(), MIGO_GAMES_DIR + "/" + gameId);
        this.cacheDir = cacheBase;
        // A directory level rather than a name suffix, so a sweep of abandoned
        // temp directories is confined to one subtree and cannot reach the
        // subpackage install store or the staging directories beside it.
        this.tempRoot = new File(cacheBase, TEMP_DIR);
        this.tempDir = new File(tempRoot, Integer.toString(sessionId));
    }

    /**
     * Validate a game ID.
     *
     * <p>Lower case only, and never a reserved device name. The id becomes one
     * path component per game and that component is the storage isolation
     * boundary, so it has to mean the same thing on a case-insensitive
     * filesystem as it does here: {@code PuzzleQuest} and {@code puzzlequest}
     * are two games on Android and one directory on Windows.
     *
     * @param gameId The game ID to validate
     * @return true if valid
     */
    public static boolean isValidGameId(String gameId) {
        return gameId != null
                && GAME_ID_PATTERN.matcher(gameId).matches()
                && !RESERVED_GAME_IDS.contains(gameId);
    }

    /**
     * Get the game ID.
     */
    public String getGameId() {
        return gameId;
    }

    /**
     * Get the user data directory (persistent).
     * <p>
     * Use for saves, preferences, user-generated content.
     * Maps to virtual path: /user
     */
    public File getUserDataDir() {
        return userDataDir;
    }

    /**
     * Get the cache directory.
     * <p>
     * Use for downloaded resources, decoded images, etc.
     * May be cleared by the system when storage is low.
     * <p>
     * This is the per-game cache root, and the game does not see it directly:
     * it also holds runtime state — the subpackage install store and record, and
     * install staging directories — so the virtual path {@code /cache} is mapped
     * to a dedicated subdirectory of it instead. A game that could write the
     * install record would decide which package bytes a later session mounts as
     * {@code /code}.
     */
    public File getCacheDir() {
        return cacheDir;
    }

    /**
     * Get the code directory (read-only for game).
     * <p>
     * Contains the game's JavaScript code and assets.
     * After the first successful signed launch, Migo clears Unix write bits on
     * the active tree. A trusted installer must stage a complete sibling tree
     * for an update; it must serialize deployment against session start, stop
     * every session for this game, and publish only a complete tree. It must not
     * overwrite the active tree in place or assume {@link File#renameTo(File)}
     * atomically replaces a non-empty directory on API 26. Use a recoverable
     * whole-tree transaction and finish recovery before the next start. A
     * detached old tree may have its write bits restored before cleanup.
     * Maps to virtual path: /code
     */
    public File getCodeDir() {
        return codeDir;
    }

    /**
     * Get the temporary directory.
     * <p>
     * Private to this session and cleared when it ends, so a second live session
     * of the same game neither reads these files nor loses them.
     * Maps to virtual path: /tmp
     */
    public File getTempDir() {
        return tempDir;
    }

    /**
     * Ensure all directories exist.
     * Call this before starting the game.
     */
    public void ensureDirectories() {
        userDataDir.mkdirs();
        cacheDir.mkdirs();
        tempDir.mkdirs();
        // codeDir is created by download/extraction logic
    }

    /**
     * Clean up this session's temporary files.
     * Call this when the session ends.
     * <p>
     * Scoped to this session's directory: it used to be the whole per-game temp
     * directory, which a second live session of the same game was still writing.
     */
    public void cleanupTemp() {
        deleteRecursive(tempDir);
        tempDir.mkdirs();
    }

    /**
     * Remove the temp directories of sessions that are no longer live.
     *
     * <p>A session's temp directory is removed by its own teardown, so what is
     * left here belongs to a session whose process died before it could run —
     * app kills are routine on Android — or to a build that wrote {@code /tmp}
     * before it was split per session, whose files sit directly in this
     * directory. Neither has an owner that will ever come back for it.
     *
     * <p>Called before this session creates its own directory, which is also what
     * makes it the guarantee that {@code /tmp} starts empty: this session's id is
     * not live yet, so a directory left behind by a dead session that happened to
     * hold the same id is swept as abandoned rather than inherited. Session ids
     * are a per-process counter, so after a restart they are handed out again from
     * the beginning and that collision is the common case, not a remote one.
     *
     * <p>{@code isLive} decides ownership rather than a timestamp or a name
     * pattern, because only liveness distinguishes "abandoned" from "in use by
     * someone else right now". Passing it in keeps this class free of the session
     * registry and lets the sweep be tested without one.
     *
     * @param isLive whether a session id still has a live session behind it
     */
    void sweepAbandonedTemp(IntPredicate isLive) {
        File[] entries = tempRoot.listFiles();
        if (entries == null) return;
        for (File entry : entries) {
            Integer owner = temporaryDirectoryOwner(entry);
            if (owner != null && isLive.test(owner)) continue;
            deleteRecursive(entry);
        }
    }

    /**
     * The session id {@code entry} belongs to, or {@code null} if it belongs to no
     * session at all — a loose file from before {@code /tmp} was split per session,
     * or a name this class did not write.
     */
    private static Integer temporaryDirectoryOwner(File entry) {
        if (!entry.isDirectory()) return null;
        try {
            return Integer.valueOf(entry.getName());
        } catch (NumberFormatException notASessionDirectory) {
            return null;
        }
    }

    /**
     * Delete all game data (uninstall).
     * <p>
     * Removes both persistent and cache data.
     */
    public void deleteAll() {
        deleteRecursive(userDataDir.getParentFile()); // Delete filesDir/.../gameId/
        deleteRecursive(cacheDir);                     // Delete cacheDir/.../gameId/
    }

    /**
     * Get the total size of user data in bytes.
     */
    public long getUserDataSize() {
        return getDirectorySize(userDataDir);
    }

    /**
     * Get the total size of cache in bytes.
     */
    public long getCacheSize() {
        return getDirectorySize(cacheDir);
    }

    private static void deleteRecursive(File file) {
        if (file == null || !file.exists()) return;
        if (file.isDirectory()) {
            // Signed launch seals the active code tree. Trusted uninstall must
            // restore owner traversal/mutation bits before removing children.
            file.setWritable(true, true);
            file.setExecutable(true, true);
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursive(child);
                }
            }
        }
        file.delete();
    }

    private static long getDirectorySize(File dir) {
        if (dir == null || !dir.exists()) return 0;
        if (dir.isFile()) return dir.length();

        long size = 0;
        File[] children = dir.listFiles();
        if (children != null) {
            for (File child : children) {
                size += getDirectorySize(child);
            }
        }
        return size;
    }

    @Override
    public String toString() {
        return "GamePaths{" +
                "gameId='" + gameId + '\'' +
                '}';
    }
}
